ml-proto/spec/int.ml

```diff
@@ -208,7 +208,55 @@ struct
   let ge_s x y = x >= y
   let ge_u x y = cmp_u x (>=) y
 
-  let of_string = Rep.of_string
+  (* This implementation allows leading signs and unsigned values *)
+  let of_string x =
+    let len = String.length x in
+    let power_of_two n = Rep.shift_left Rep.one n in
+    let ten = Rep.of_int 10 in
+    let fail () = failwith "int_of_string" in
+    let parse_hexdigit c =
+      int_of_char c -
+        if '0' <= c && '9' >= c then 0x30
+        else if 'a' <= c && 'f' >= c then 0x61 - 10
+        else if 'A' <= c && 'F' >= c then 0x41 - 10
+        else fail ()
+    in
+    let parse_hex offset sign =
+      let num = ref Rep.zero in
```

Comment:
Well, it should be tail-recursive, of course. Like so:
```
let rec parse_hex i num =
  if i = len then num
  else if overflow... then fail ()
  else parse_hex (i+1) (logor (shift_left num 4) (of_int (parse_hexdigit x.[i])))
```
Btw, I think you can get rid of the entire sign-specific logic in these functions. For values with negative sign it should suffice to simply check once in the end (at line 257 below) whether the value is negative already.